### ISSUE TITLE
Bake timeout settings

### ISSUE THREAD

**Issue by HariSekhon** [spinnaker/spinnaker#6448]
### Issue Summary:

Bake stage fails with timeout after 20 seconds

### Cloud Provider(s):

GCP

### Environment:

Spinnaker 1.26.3 deployed fully distributed on kubernetes (GKE 1.17) via Halyard 1.42.1

### Feature Area:

Pipelines

### Description:

Bake stage fails with timeout after 20 seconds when trying to clone a large GitHub repository containing the kubernetes manifests on which to run a Kustomize step

### Steps to Reproduce:

Large GitHub repo used as manifest artifact source for kubernetes manifests used by pipeline's Bake stage doing Kustomize

### Additional Details:

I've been scouring the documentation but can't find timeout settings, this page is closest but doesn't appear to have a complete list of settings?

https://spinnaker.io/reference/halyard/custom/#tweakable-service-settings

I also looked through past tickets such as #5518 and #4498, but the timeout settings shown there have not worked for me when putting them under `~/.hal/default/service-settings/` and then running `hal deploy apply`. I've tried both `orca.yml` / `rosco.yml` as well as `orca-local.yml` / `rosco-local.yml` and sections `ok-http-client` and `okHttpClient` as mentioned in those tickets.

Could we please get clearly documented settings to tune timeouts?

**Comment by HariSekhon** [comment]
This is one of our larger Git repos, it takes 27 secs to check out from scratch:

```
time git clone "https://$GITHUB_TOKEN@github.com/$org/$repo"
Cloning into '$repo'...
remote: Enumerating objects: 201362, done.
remote: Counting objects: 100% (1826/1826), done.
remote: Compressing objects: 100% (808/808), done.
remote: Total 201362 (delta 1161), reused 1490 (delta 985), pack-reused 199536
Receiving objects: 100% (201362/201362), 127.79 MiB | 5.67 MiB/s, done.
Resolving deltas: 100% (146179/146179), done.
Checking out files: 100% (6584/6584), done.

real    0m27.685s
user    0m4.927s
sys     0m3.705s
```

On top of that, Devs are pushing to consolidate some of our other repos into this one to form a Monorepo!

**Comment by HariSekhon** [comment]
Also, is it possible to use a shallow clone, as this is faster, and I've noticed that several CI/CD systems do this by default?

For example, on this repo it takes the checkout from 27 seconds down to 12 seconds:

```
time git clone --depth 1 "https://$GITHUB_TOKEN@github.com/$org/$repo" 
Cloning into '$repo'...
remote: Enumerating objects: 8373, done.
remote: Counting objects: 100% (8373/8373), done.
remote: Compressing objects: 100% (7125/7125), done.
remote: Total 8373 (delta 2168), reused 3997 (delta 741), pack-reused 0
Receiving objects: 100% (8373/8373), 40.28 MiB | 4.71 MiB/s, done.
Resolving deltas: 100% (2168/2168), done.
Checking out files: 100% (6584/6584), done.

real    0m12.828s
user    0m1.534s
sys     0m1.901s
```

**Comment by HariSekhon** [comment]
Since the Bake stage doesn't have the option to `Fail stage after a specific amount of time` in the Spinnaker UI, I copied the JSON field from the Deploy stage and added this to the JSON directly for the Bake stage, which was persisted without error but has made no difference:
```
  "stageTimeoutMs": 600000,
```

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by shiwano** [comment]
@spinnakerbot remove-label stale

We really need this!

**Comment by HariSekhon** [comment]
Spinnaker was dead in the water for us due to this issue of not being able to handle a non-trivial repo checkout, we went with ArgoCD instead.

**Comment by achew22** [comment]
I added this additional configuration to `local/orca-local.yml` and `local/rosco-local.yml` when using github.com/spinnaker/kustomization-base as my configuration for the project and it's been working great for a couple of months. This is for the next person who runs into this, best of luck with ArgoCD.

```
# Add a timeout to okHttp so that it has 120s to clone repos which tend to
# time out with their default of 20s
# https://github.com/spinnaker/rosco/pull/594#pullrequestreview-395576692
okHttpClient:
    readTimeoutMs: 120000
```

**Comment by HariSekhon** [comment]
I think I tried these at the time, iirc it depended on versions:
```
$ cat > ~/.hal/default/service-settings/orca.yml <<EOF
okHttpClient:
  connectTimeoutMs: 1200000
  readTimeoutMs: 1200000
EOF

$ cat > ~/.hal/default/service-settings/rosco.yml <<EOF
okHttpClient:
  connectTimeoutMs: 1200000
  readTimeoutMs: 1200000
EOF

$ hal deploy apply
```

```
$ cat > ~/.hal/default/service-settings/orca.yml <<EOF
ok-http-client:
  connectTimeoutMs: 1200000
  readTimeoutMs: 1200000
EOF

$ cat > ~/.hal/default/service-settings/rosco.yml <<EOF
ok-http-client:
  connectTimeoutMs: 1200000
  readTimeoutMs: 1200000
EOF

$ hal deploy apply
```

but I don't recall either of them working to prevent the bake timeout.

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by shiwano** [comment]
@spinnakerbot remove-label to-be-closed

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by spinnakerbot** [comment]
This issue is tagged as 'to-be-closed' and hasn't been updated  in 45 days, so we are closing it. You can always reopen this issue if needed.

**Comment by brantburnett** [comment]
> I added this additional configuration to `local/orca-local.yml` and `local/rosco-local.yml` when using github.com/spinnaker/kustomization-base as my configuration for the project and it's been working great for a couple of months. This is for the next person who runs into this, best of luck with ArgoCD.
> 
> ```
> # Add a timeout to okHttp so that it has 120s to clone repos which tend to
> # time out with their default of 20s
> # https://github.com/spinnaker/rosco/pull/594#pullrequestreview-395576692
> okHttpClient:
>     readTimeoutMs: 120000
> ```

This fix worked for me as well, though I did 60 seconds because I didn't need 120. I do feel like this should either be the default, be more easily configurable, and/or handled through some means other than a long-running HTTP request.

**Comment by Hillkorn** [comment]
None of the above did work for me because the issue is not the client but the server. After some debugging I found out that the server's async request handling has timeouts as well and the default is 30 seconds. This is a problem for slow git clone jobs that are done in that async requests. To fix this I had to increase the timeout with the following config in `.hal/default/service-settings/clouddriver.yml`
```
env:
  JAVA_OPTS: "-XX:MaxRAMPercentage=50.0 -Dspring.mvc.async.request-timeout=180000"
```
